### PR TITLE
Fixed the missing debug menus to visualize the ray tracing acceleration structure (case 1371410).

### PR DESCRIPTION
**Pull request by anisunity**
https://fogbugz.unity3d.com/f/cases/1371410/
This PR adds a full screen debug menu that allows the user to visualize what is in the various layers of the acceleration structure (and to understand if an object is missing). This allowed to figure out that a bug was introduced with the c++ culling feature that needs to be fixed in a separate (engine PR).
This also allowed to remove the flagging of the shadow flags when the light had ray traced shadows but disabled shadows.

The user has now two sub options to pick in what mode they want to display (primitiveID, instanceID) and view (reflection, GI, Path tracing, etc)
image

Testing status
Tried to test as much as possible all the different combinations

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/HDShaderPasses.cs [hunk 1]
```diff
@@ -1286,6 +1286,61 @@ IncludeCollection GenerateIncludes()
...
 
+        #region Raytracing Debug
+
+        public static PassDescriptor GenerateRaytracingDebug()
+        {
+            return new PassDescriptor
+            {
+                // Definition
+                displayName = "DebugDXR",
+                referenceName = "SHADERPASS_RAYTRACING_Debug",
```
Comment (arttu-peltonen) [inline com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/HDShaderPasses.cs:1297]: Debug -> uppercase?

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/ShaderPass.cs.hlsl [hunk 2]
```diff
@@ -31,8 +31,9 @@
 #define SHADERPASS_RAYTRACING_GBUFFER (21)
 #define SHADERPASS_RAYTRACING_SUB_SURFACE (22)
 #define SHADERPASS_PATH_TRACING (23)
-#define SHADERPASS_CONSTANT (24)
-#define SHADERPASS_FULL_SCREEN_DEBUG (25)
+#define SHADERPASS_RAY_TRACING_DEBUG (24)
```
Comment (arttu-peltonen) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/ShaderPass.cs.hlsl:34]: Does this define work? Elsewhere you're using SHADERPASS_RAYTRACING_DEBUG (without underscore in the middle).

File: com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs [hunk 3]
```diff
@@ -387,6 +413,12 @@ internal DebugDisplaySettings()
                 .ToArray();
             s_MsaaSamplesDebugValues = (int[])Enum.GetValues(typeof(MSAASamples));
 
+            s_RTASViewDebugStrings = Enum.GetNames(typeof(RTASDebugView)).Select(t => new GUIContent(t)).ToArray();
```
Comment (arttu-peltonen) [inline com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs:416]: I think you can get rid of these arrays by using DebugUI.EnumField.autoEnum.

looks like the MSAADebugValues/Strings properties are never even used. Fixing optional :)

**Review by arttu-peltonen — CHANGES_REQUESTED** [review]
Looks like a super useful debug mode if you work with RT! I mostly reviewed the DebugDisplay integration which looks good. I'm less familiar with the shader stuff, but I did notice something that looks suspicious with keywords (or at least doesn't conform to style), so please double-check that.

File: com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/HDTarget.cs [hunk 4]
```diff
@@ -1535,6 +1541,15 @@ static class CoreKeywordDescriptors
...
         };
+
+        public static KeywordDescriptor RayTracingDebug = new KeywordDescriptor()
+        {
+            displayName = "RayTracingDebug",
+            referenceName = "RAY_TRACING_DEBUG",
+            type = KeywordType.Boolean,
+            definition = KeywordDefinition.MultiCompile,
+            scope = KeywordScope.Global
+        };
```
[the remark this answers is not in the document]
Comment (anisunity) [reply com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/HDTarget.cs:1552]: Yeah, they debug_display keyword may be triggered and is not used in the same context

File: com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/HDShaderPasses.cs (hunk 1 above)
Comment (anisunity) [reply com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/HDShaderPasses.cs:1297]: Indeed

File: com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs (hunk 3 above)
Comment (anisunity) [reply com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs:416]: It's automatically generated, (RayTracing => RAY_TRACING, raytracing => RAYTRACING). Should be with a T upper case

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/ShaderPass.cs.hlsl (hunk 2 above)
Comment (anisunity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/ShaderPass.cs.hlsl:34]: It's automatically generated, (RayTracing => RAY_TRACING, raytracing => RAYTRACING). Should be with a T upper case

File: com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs (hunk 3 above)
Comment (anisunity) [reply com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs:416]: image

Comment (anisunity) [reply com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs:416]: it seems to be used

Comment (arttu-peltonen) [reply com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs:416]: yeah sorry I meant only written to, never read.

Comment (arttu-peltonen) [reply com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs:416]: When you use autoEnum, you don't need to provide the enumValues and enumNames when constructing the DebugUI.EnumField. So you only want to manually pass in the values/names if you want a subset of the enum or something. But it's really optional to fix, just FYI.

Comment (anisunity) [reply com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.cs:416]: Yeah, I am too lazy to fix this one. I'll fix the other ones

**Review by remi-chapelain — APPROVED** [review]
I made public (with added summary) some function to be able to fetch this new debug mode from the test framework which I also modified. 

3 tests have been added to prevent future regression. 
- One to test instanceID with RTR
- One to test primitiveID with RTR
- One to test primitiveID with RTShadows

The only problem is that instanceID is not deterministic between scene launch, so there is only one object in the instanceID scene to prevent randomness instanceID attribution. 

**_2009_Debug_RTAS_ScreenSpaceReflections_InstanceID_**
2009_Debug_RTAS_ScreenSpaceReflections_InstanceID

**_2009_Debug_RTAS_Shadows_PrimitiveID_**
2009_Debug_RTAS_Shadows_PrimitiveID

**_2009_Debug_RTAS_ScreenSpaceReflections_PrimitiveID_**
2009_Debug_RTAS_ScreenSpaceReflections_PrimitiveID

Other than that, tested the feature with the rest of the effects / manual RTAS and light layers. ✔️

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingIntersection.hlsl [hunk 5]
```diff
@@ -11,11 +11,37 @@ struct RayCone
     float spreadAngle;
 };
 
+// Structure that defines the current state of the visibility
+struct RayIntersectionDebug
+{
+    // Distance of the intersection
+    float t;
+    // Barycentrics of the intersection
+    float2 barycentrics;
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingIntersection.hlsl:20]: This doesn't seems to be used, is it useful? barycentrics

**Review by sebastienlagarde — COMMENTED** [review]
Let's discuss. From what I read in the code you don't need to add a SHADER_RAYTRACING_Debug  pass to all shader, you can only have a generic one. Is there any reasons you chose to not do that?

**Review by arttu-peltonen — APPROVED** [review]

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingIntersection.hlsl (hunk 5 above)
Comment (anisunity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingIntersection.hlsl:20]: Yes, it's for the future, a debug mode that we may want to use

**Comment by anisunity** [comment]
As discussed we cannot have a generic closest hit in the ray traced, so i'll be trying the shaderlab fallback feature see if it helps

**Comment by anisunity** [comment]
@remi-chapelain you forgot to include the materials in your commit :(

**Comment by anisunity** [comment]
Something is currently broken with the test framework, the tests wont run when  launching all of em.
The tests that have been added in this PR are green...
image

**Review by cdxntchou — APPROVED** [review]
Looks good.  I think Terrain is defining the same usepass list, but you guys look closer to landing, so we can work with them to adapt their PR.

**Review by sebastienlagarde — APPROVED** [review]

**Pull request merged**